### PR TITLE
Sub-package install

### PR DESCRIPTION
**Pull request by sir-sigurd**
## Description

This allows to install only the part of the package. Example:

We have package `foo/bar`:
```
(remote Package)
 └─requirements.txt
 └─scripts/
   └─build.py
   └─demolish.py
```

```python
In [17]: import os, tempfile, quilt3                 

In [18]: with tempfile.TemporaryDirectory() as dest: 
    ...:     quilt3.Package.install('foo/bar/scripts', dest=dest) 
    ...:     print([os.path.relpath(os.path.join(top, f), dest) for top, dirs, files in os.walk(dest) for f in files]) 

['demolish.py', 'build.py', 'more_scripts/test.py']
```

```python
In [20]: with tempfile.TemporaryDirectory() as dest: 
    ...:     quilt3.Package.install('foo/bar/scripts/build.py', dest=dest) 
    ...:     print([os.path.relpath(os.path.join(top, f), dest) for top, dirs, files in os.walk(dest) for f in files]) 

['build.py']
```

### CODE REVIEW

**Comment by akarve** [comment]
@sir-sigurd looks like we have some unit tests in `test_packages` failing

**Comment by akarve** [comment]
I took a look and I _think_ at least one failure is because we are stubbing S3 calls that are never made.

**Comment by mistercrunch** [comment]
# [Codecov](https://codecov.io/gh/quiltdata/quilt/pull/1541?src=pr&el=h1) Report
> Merging [#1541](https://codecov.io/gh/quiltdata/quilt/pull/1541?src=pr&el=desc) into [master](https://codecov.io/gh/quiltdata/quilt/commit/f046f8366c939edf791bd4454ce46c6db5b8422d&el=desc) will **increase** coverage by `0.39%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/quiltdata/quilt/pull/1541?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master    #1541      +/-   ##
==========================================
+ Coverage   83.33%   83.73%   +0.39%     
==========================================
  Files          27       27              
  Lines        3715     3769      +54     
==========================================
+ Hits         3096     3156      +60     
+ Misses        619      613       -6     
```


| [Impacted Files](https://codecov.io/gh/quiltdata/quilt/pull/1541?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [quilt3/search\_util.py](https://codecov.io/gh/quiltdata/quilt/pull/1541/diff?src=pr&el=tree#diff-cXVpbHQzL3NlYXJjaF91dGlsLnB5) | `75.86% <0.00%> (-3.45%)` | :arrow_down: |
| [tests/integration/test\_packages.py](https://codecov.io/gh/quiltdata/quilt/pull/1541/diff?src=pr&el=tree#diff-dGVzdHMvaW50ZWdyYXRpb24vdGVzdF9wYWNrYWdlcy5weQ==) | `100.00% <0.00%> (ø)` | :arrow_up: |
| [tests/conftest.py](https://codecov.io/gh/quiltdata/quilt/pull/1541/diff?src=pr&el=tree#diff-dGVzdHMvY29uZnRlc3QucHk=) | `92.30% <0.00%> (+0.87%)` | :arrow_up: |
| [quilt3/packages.py](https://codecov.io/gh/quiltdata/quilt/pull/1541/diff?src=pr&el=tree#diff-cXVpbHQzL3BhY2thZ2VzLnB5) | `90.39% <0.00%> (+1.20%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/quiltdata/quilt/pull/1541?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/quiltdata/quilt/pull/1541?src=pr&el=footer). Last update [f046f83...f479057](https://codecov.io/gh/quiltdata/quilt/pull/1541?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: api/python/quilt3/packages.py [hunk 1]
```diff
@@ -436,12 +438,26 @@ def install(cls, name, registry=None, top_hash=None, dest=None, dest_registry=No
                     f"'build' instead."
                 )
 
+        subpkg_key_index = name.find('/', name.find('/') + 1)
```
Comment (kevinemoore) [inline api/python/quilt3/packages.py]: Searching for '/' in the string is a bit confusing and seems a bit fragile. Can you use a regex here? Preferably update the `quilt3.util.PACKAGE_NAME_FORMAT` to allow an optional sub-package path. You should be able to use that regex to separate the package name and sub-package path.

Comment (sir-sigurd) [reply api/python/quilt3/packages.py]: > Preferably update the `quilt3.util.PACKAGE_NAME_FORMAT` to allow an optional sub-package path. You should be able to use that regex to separate the package name and sub-package path.

I think we should add the separate regex as we don't allow sub-package in other places, makes sense?

Comment (akarve) [reply api/python/quilt3/packages.py]: Yes, please proceed with the separate regex. You can create e.g. `quilt3.util.SUBPACKAGE_NAME_FORMAT`.

**Comment by sir-sigurd** [comment]
I deleted my fork repo and now I have to open new PR to modify it :face_with_head_bandage: .
New PR: https://github.com/quiltdata/quilt/pull/1549

**Pull request closed without merging**